### DISCUSSION TITLE
[Question] Migrating to arbitrary version

### DISCUSSION THREAD

**Discussion by kshpytsya** [sqlalchemy/alembic#1214] [Ideas]
In our project, the "D" in "CI/CD" stands for "delivery". The build pipeline produces Docker images which, per [this](https://github.com/sqlalchemy/alembic/issues/742#issuecomment-700932424) reasonable suggestion also include Alembic migrations. Deployment is managed in GitOps-y way with a dedicated repository describing which image versions are deployed to which environments. It is my intent to perform DB migrations as one of the steps in deployment pipeline. While migrations in production environment are expected to be linear upgrades, for our dev/sandbox environment I expect situations with having to first deploy image version containing "A->B" (let's call it image `v1.feat1`) and then that containing "A->C" (lets call it image `v1.feat2`) migrations. The problem here is that `v1.feat2` knows nothing of migration `B` and `v1.feat1` knows nothing of migration `C`, making it impossible to use either of those two images to perform migrations necessary to switch between these two images. Therefore one possible solution I see is the following three-step procedure:

* Alembic in the image containing the version to be deployed ("new") is queried to report entire migration version graph known to it and the head migration version.
* Alembic in the image containing already deployed version ("old") is supplied with information from the first step and is requested to perform all the migrations it needs to reach some version that is known to the "new" image.
* If the "old" image doesn't know of the head migration for the "new" image, alembic in the "new" image is requested to perform remaining migrations.

The drawback of this solution is that migrations are not going to be performed in a single DB transaction. Other possible solutions are:

* Two steps:
  * The "old" image is queried to export all the migrations in some Alembic-specific format
  * The "new" image is supplied the file produced by the previous step thus making it capable of dealing with current DB versions it wouldn't normally know of.
* `alembic/versions/*.py` files are copied into the same directory of the new image.

I am not sure whether these alternative solutions would not have any issues with different versions of Alembic (or some other library, or even version of Python) in the "old" and "new" images. These solutions have the benefit of performing all migration steps in a single DB transaction.

Of course, regardless of the approach taken, Alembic would still need a "deploy" (tentative name) command instead of or in addition to "upgrade"/"downgrade".

While targeted at dev environment, the same mechanism would also work for downgrades in production (which are hoped to be exceptional situations).

This may be already a solved problem, but questions like [this](https://stackoverflow.com/questions/37035260/is-there-a-simple-way-to-tell-alembic-to-migrate-to-a-specific-revision) make me doubt it, and I would also expect to find something in the [cookbook](https://alembic.sqlalchemy.org/en/latest/cookbook.html).

**Comment by zzzeek** [comment]
the stack overflow question refers to someone who doesn't know if their current migration target is an upgrade or a downgrade, relative to the current state of the database.  they of course should look at "alembic history" to see where the database's current version is relative to history, but downgrades are not really a normal part of automated deployments and they are very much separate from "upgrades".   your issue seems to be related to deployment, so I dont see how that question applies here.

if you have an A->B A->C kind of problem, those sound like branches.     if two branches of code have diverging alembic trees, you first merge them into a common merge point and alembic can move to that point.   I would check out https://alembic.sqlalchemy.org/en/latest/branches.html

**Comment by kshpytsya** [comment]
I hoped it would be obvious from my description, that in my situation, deployment pipeline, which I intend to be the place where alembic migrations are executed, doesn't know whether it is "upgrade" or "downgrade". It just knows the version of our application it needs to deploy and also the version that is currently deployed. So I am looking for a way to be able to apply a necessary sequence of migrations that would bring the database from "old" to "new" version. One of the situations is the need to switch from "A->B" to "A->C", which would involve "downgrade B->A" and then "upgrade A->C". As an extra complication, one image, containing alembic knows only of ["A", "B"] and another one knows only of ["A", "C"].

For prod I mostly expect regular upgrades, however, there is a possibility that we would need to do a roll-back to previous version which would imply doing downgrade. But for dev we need to be able to conveniently switch our single dev cluster between different branches of our app, which would also possibly involve abovementioned db migrations.

[This](https://github.com/sqlalchemy/alembic/discussions/1161) seems to be remotely related.

**Reply by zzzeek** [reply]
OK so...you need to run one operation and have it do either upgrade or downgrade?    as mentioned, use the output of alembic history, which you can also link into programmatically.  I would look at the source code to the command and see how to programmatically walk the whole tree and compare it to whatever version the database is at:

https://github.com/sqlalchemy/alembic/blob/3d9b1128cd6bf03ecb45003587c0eedfb9552b07/alembic/command.py#L452-L519

alembic's API within there that's not underscored is public so I would look through to get the bits that you need.

**Reply by kshpytsya** [reply]
Do I understand it correctly that you consider this to be outside the scope of Alembic itself, despite me being not the only one with such a request? I just would like to know official position before starting to implement my own bicycle.

**Reply by zzzeek** [reply]
guessing for a certain ID if we should use upgrade or downgrade automatically, I think that's out of scope yes.

**Reply by CaselIT** [reply]
> Do I understand it correctly that you consider this to be outside the scope of Alembic itself, despite me being not the only one with such a request? I just would like to know official position before starting to implement my own bicycle.

you mean something like `alembic move_to arbitraty_id`? where arbitrary it may be in the after the current revision (so upgrade) or before it (so downgrade)?

**Comment by kshpytsya** [comment]
Digging through Alembic issues found this one:

https://github.com/sqlalchemy/alembic/issues/464#issuecomment-734648363

so it seems like this is a recurring idea. It's a shame that it doesn't get traction with Alembic's authors.

P.S. Looking at [this](https://github.com/sqlalchemy/alembic/pull/790) PR probably gives an idea of complexity of proper implementation of this feature.

**Reply by CaselIT** [reply]
I'm not sure how that would solve this part contraints

> As an extra complication, one image, containing alembic knows only of ["A", "B"] and another one knows only of ["A", "C"].

**Reply by kshpytsya** [reply]
> I'm not sure how that would solve this part contraints

As I have mentioned in my original question, I envision that the tool that implements this (be it the alembic itself or some wrapper) would implement it as a three or two phase process (I have described pros and cons for both), so it would be invoked as part of deployment process as (for 3-phase):

```sh
docker run -v/tmp/dir:/tmp/dir <new-image> tool migrate --phase1 -o /tmp/dir/phase1.json
docker run -v/tmp/dir:/tmp/dir:ro <old-image> tool migrate --phase2 -i /tmp/dir/phase1.json
docker run <new-image> tool migrate --phase3
```

or, for 2-phase:

```sh
docker run -v/tmp/dir:/tmp/dir <old-image> tool export-migrations -o /tmp/dir/migrations.json
docker run -v/tmp/dir:/tmp/dir:ro <new-image> tool migrate --import-migrations /tmp/dir/migrations.json
```

Command and parameter names are tentative. The revision that we want to migrate to is "head" of the "new-image".

**Reply by CaselIT** [reply]
do you know the common migration to old image and new image, A in your examples?

Because if you do, can't you use the following?
```sh
# in old image
alembic downgrade A
```
```sh
# in new image
alembic upgrade head
```

**Reply by kshpytsya** [reply]
No, I do not know the common migration, it is the task of deployment process to figure it out automatically.

**Reply by zzzeek** [reply]
re: issue #464, that was a very complicated add, and that submitter actually contributed the code and spent many weeks collaborating on its implementation.

As I mentioned previously, given a single identifier and determining if it's "up" or "down" based on the current version in the alembic_version table is relatively trivial.

If, OTOH, the feature request, which I had a lot of trouble understanding from the description, is that some database is at some version that is **both** "up" and "down" at the same time relative to some other alembic version tree that represents a later version, that feature is not possible; the migration files themselves would not be compatible with such a database and that idea is a non-starter.    Since you stated you think this problem is of similar complexity to #464 makes me think that might be what you're thinking of.

**Reply by kshpytsya** [reply]
I am totally confused as to the `both "up" and "down" at the same time` part, which I attribute to my limited understanding of vast possibilities of Alembic (ab)use.

**Reply by kshpytsya** [reply]
Just to give a broad picture: I want to have a system that "Just works"™, meaning that developers that do not change any db schemas in the code should not know anything about migrations at all, those devs that change db schemas should know how to create proper alembic migrations (probably with restriction on having a single head) locally (there will be a local instance of the same db engine running on dev's machine), and only if something goes wrong, a specially trained person should do the investigation as to why and how. Deploying arbitrary versions of the application should not become more complicated due to performing db migrations being one of the steps done by deployment pipeline. Of course, there is a possibility of data loss but that's what continuous DB backups are for.

**Reply by zzzeek** [reply]
alembic already does all those things.    I dont understand the part where someone has version 1.1 of the app, then downgrades to an older version of it, and somehow does not know if they need to type "upgrade" or "downgrade" .       if developers are working on branches that have divergent db migration files, then by definition they are devs that "change db schemas" and they have to create merge points.

**Reply by zzzeek** [reply]
Alembic's branching model is described at https://alembic.sqlalchemy.org/en/latest/branches.html .  As you noted, in #464 a contributor came in to improve some aspects of this, however that developer was able to provide us with a very specific scenario, an exact branch structure and the exact operation they wanted.    After that, I still did not have the resources to provide this feature but I worked with them on getting it completed.

So far the descriptions of what is desired here have been quite vague.    I would suggest you dig into specific scenarios that are not working and determine what exactly needs to happen first in order to satisfy those scenarios.

I would also note that alembic is not the only migration tool out there.   Django's tool is much more automated and generally is perceived as very easy to use.  Alembic is not.  I am actually fine with that, so you might want to look into what they do as it is pretty much the "push button" experience you seem to be looking for.

**Comment by kshpytsya** [comment]
> . I dont understand the part where someone has version 1.1 of the app, then downgrades to an older version of it, and somehow does not know if they need to type "upgrade" or "downgrade" . if developers are working on branches that have divergent db migration files, then by definition they are devs that "change db schemas" and they have to create merge points.

The thing is that deployments of versions that introduce schema changes may not always be done by people developing those feature branches.

Let me remind that we have a single shared dev environment where all the devs can deploy. They do not step on each others toes by communicating via messenger. Consider the following scenario:

* Alice develops a feature branch that includes changes to DB. Eventually she deploys it to check it on real system or to let other developer Charlie that work on other apps to try out new features. Alice is done for the day, keeping her feature branch deployed.
* Bob is working on some other feature (that doesn't change anything in DB) and wants to deploy. He asks whether anyone is using the app on dev, and deploys his version. This causes an alembic downgrade of which Bob knows nothing about.
* Charlie needs to interact with Alice's branch again, and, after making sure it is ok to deploy, deploys Alice's version (or asks Bob, as Charlie may even not know how to deploy the app). This causes alembic upgrade of which neither Charlie nor Bob know about.

Similar scenario can be with A, B and C db migrations and a need to switch between feature branches containing B and C (and possibly back). And that switch can be done by Daniel, who just knows of feature branches and knows nothing of migrations. He just knows how to activate the deployment pipeline specifying desired app version.

**Reply by zzzeek** [reply]
> Bob is working on some other feature (that doesn't change anything in DB) and wants to deploy. He asks whether anyone is using the app on dev, and deploys his version. This causes an alembic downgrade of which Bob knows nothing about.

this is a very strange collaboration pattern.  you have one single shared server, with developers who dont know how the software works with the database, deploying themselves to that single shared server.    This is not a good working model.  If  Alice puts a newer version of the software on the enviroment, then Bob comes and wants to run an older version, he would have to communicate with Alice where Alice would confirm that Bob first needs to **downgrade** the Alice version of the application to the common version, then install the **older** software that lacks the Alice version.  

There are many collaboration patterns that are vastly superior to this one:

1. everyone has their own development database.  This is the most common. It's 2023, we no longer need DBA-installed DBs on $30K hardware just to do development.
2. there are multiple dev databases dedicated to the working models for different developers.  These can still be changed, however when Alice no longer needs to work on a particular database, that DB can be wiped out and built from scratch.   
3. Each developer who wants to run some experiemental feature branch on the "single dev DB" is then responsible for reverting it to a common known state when that work is complete.

There's no option to have the application with Alices "feature branch" version present, then Bob's "does not have the feature branch" version of the software is installed, then Bob's version of the alembic migrations "downgrades Alices changes". That's not possible because Bob's version of the software does not contain the downgrade steps.  Only Alice's does.  if only Alice knows about how to upgrade/downgrade these new steps, then she is responsible for un-doing them when they are no longer needed.

**Reply by kshpytsya** [reply]
As for the impossibility of downgrade of Alice's version by Bob: this is what the three or two phase process I have described in my original question is intended to solve. The deployment system knows the app version currently deployed (Alice's) and the one that needs to be deployed (Bob's). So, those two docker images combined contain all the alembic migration files needed to migrate between those versions.

I appreciate your elaborate comment on collaboration patterns but they are not mine to change in this project, I need to make do with whatever process is given, and I am sure that technically it is possible.

**Reply by zzzeek** [reply]
then you are building something very custom so again I suggest using Alembic's API to pull in the current version number and make these special decisions programmatically.

**Reply by zzzeek** [reply]
also noting, if there *were* a feature to add to alembic here, I dont have the resources to implement myself, so if you work with the API directly here, that's where the beginnings of a "feature" would start in any case.    I suggest getting the thing you want working first.  Getting it to be "officially part of alembic" is secondary.  Getting someone else to actually do the dev work here, that's not happening.  I dont have development resources for unusual new use cases.

**Reply by kshpytsya** [reply]
Ok, agree to disagree, let it be a custom solution.
Now, a practical question. What would be a proper way of migrating from version "A->B A->C B->D C->D" to "A->B"? I believe it should be just "undo the changes introduced by A->C". How would I compute that set of operations and how do those operations map into Alembic's API?

**Reply by zzzeek** [reply]
if you have a version of the app that has generated this migration structure:

```
A->B         -> D
 ->C@branch1 -> D
```

D by definition is a mergepoint.   

either way, that app has a "downgrade" where you can downgrade back to:

```
   A->B
```

that's where the code to do this is.   the feature we did in #464 should allow this operation exactly, by doing "alembic downgrade branch1@A".   Now if the mergepoint gets in the way for that specific case, I'm not sure.   let me know if it does.  not sure what the solution would be as I deal with this branching stuff basically never.

**Reply by kshpytsya** [reply]
Kind thanks for explanations and for Alembic. So, my plan is to implement the desired logic myself. I will chime in here if I hit some edge cases that will cause me doubts.
